docs/design/proposals/split-repo/split-repo.md

```diff
@@ -0,0 +1,103 @@
+* Individual AWS service-specific controllers (`/services/$SERVICE`)
+* Individual AWS service-specific tests (`/test/e2e/$SERVICE`)
+
+When we generate service controller code in ACK, we build the `ack-generate`
+CLI from the code in this monorepo and output the generated code into the
+`/services/$SERVICE` directory also within this monorepo.
+
+When we build releases in ACK, the releases are Git tags on the monorepo --
+e.g. `v0.0.2` -- and then individual ACK service controller binaries (Docker
+images) are built and pushed to the [ECR public ACK controller image][ecr]
+registry with an image tag of `$SERVICE-$RELEASE_VERSION` -- e.g. `s3-v0.0.2`
+
+[ecr]: https://gallery.ecr.aws/aws-controllers-k8s/controller
+
+### Problem #1: Release series inflexibility
+
+This monorepo tightly couples the ACK common runtime code, code
+generator/templates and the individual service controller code, making the
+release process inflexible and cumbersome. The Elasticache team cannot have
+separate release series -- say, a `stable` and a `preview` release series --
+because releases are Git tags, and Git tags are specific to one source
+repository. If we had a `stable` and `preview` Git branch, all service
+controllers would need to follow the same release cadence and series naming. In
+addition, numeric releases such as `v0.2.0` refer to a specific
+content-addressable SHA1 Git commit, and with all service controllers in the
+same source repository, this means that numeric Git tags cannot applied to two
+different service controllers to represent an incremental change in one
+specific service controller's codebase.
+
+### Problem #2: Common runtime import inflexibility
+
+Furthermore, due to the code generator and ACK common runtime code being in the
+same repo as the service controllers that `import` this code, it's not easy to
+depend on a specific version of that common runtime from within the service
+controller code itself. Individual service controller package directories would
+need their own `go.mod` files and managing this would be a pain and wouldn't
+solve Problem #1 anyway.
+
+## Solution
+
+We propose splitting the monorepo into a set of source code repositories that
+can be separately Git-tagged and have their releases managed according to the
+preferences of the team that owns that component or individual ACK service
+controller.
+
+Specifically, we propose the following changes:
+
+1) All code for generating code will move into a new
+   `github.com/aws/aws-controllers-k8s-generate` source repository.
+
+   This includes all content in the existing `/cmd/ack-generate`, `/templates`,
+   `/pkg/generate` and `/pkg/model` directories of the existing
+   `github.com/aws/aws-controllers-k8s` source repository.
+
+2) The ACK common runtime and core ACK types will remain in the
+   `github.com/aws/aws-controllers-k8s` source repository.
+
+   The `/test` directory will contain only common testing framework code (both
```

Comment:
How would the current Github actions (https://github.com/aws/aws-controllers-k8s/actions?query=workflow%3A%22deploy+to+mkdocs%22) workflows
```
e2e-test
build-unit-test
cleanup self-hosted runner state
deploy to mkdocs 
```
be setup for individual service controllers?

For example, would each service controller needs to setup its own Github Actions -> workflows for automated e2e tests, self-hosted runner or can common tools/fleet setup be leveraged instead? So that such common setup is config driven and can be shared across service controllers instead of being repeated?

Reply:
> How would the current Github actions (https://github.com/aws/aws-controllers-k8s/actions?query=workflow%3A%22deploy+to+mkdocs%22) workflows
> 
> ```
> e2e-test
> build-unit-test
> cleanup self-hosted runner state
> deploy to mkdocs 
> ```
> 
> be setup for individual service controllers?
> 
> For example, would each service controller needs to setup its own Github Actions -> workflows for automated e2e tests, self-hosted runner or can common tools/fleet setup be leveraged instead? So that such common setup is config driven and can be shared across service controllers instead of being repeated?

We will continue to use the main github.com/aws/aws-controllers-k8s repository for all e2e tests and build scripts (for now) and just change the source and destination directories for the scripts to point at the new controller repositories instead of the `services/$SERVICE` directory in the main repo.